src/compression_chunk_size.c

```diff
@@ -90,3 +90,34 @@ ts_compression_chunk_size_totals()
 
 	return sizes;
 }
+
+int64
+ts_get_before_compression_row_count(int32 uncompressed_chunk_id)
+{
+	int found_cnt = 0;
+	int64 rowcnt = 0;
+	ScanIterator iterator =
+		ts_scan_iterator_create(COMPRESSION_CHUNK_SIZE, AccessShareLock, CurrentMemoryContext);
+	init_scan_by_uncompressed_chunk_id(&iterator, uncompressed_chunk_id);
+	ts_scanner_foreach(&iterator)
+	{
+		bool nulls[Natts_compression_chunk_size];
+		Datum values[Natts_compression_chunk_size];
+		bool should_free;
+		HeapTuple tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
+
+		heap_deform_tuple(tuple, ts_scan_iterator_tupledesc(&iterator), values, nulls);
+		if (!nulls[AttrNumberGetAttrOffset(Anum_compression_chunk_size_numrows_pre_compression)])
+			rowcnt = DatumGetInt64(values[AttrNumberGetAttrOffset(
+				Anum_compression_chunk_size_numrows_pre_compression)]);
+		if (should_free)
+			heap_freetuple(tuple);
+		found_cnt++;
+	}
+	if (found_cnt != 1)
+		elog(ERROR,
+			 "missing record for chunk with id %d in %s",
```

Comment:
This is a primary key.

Reply:
> Few comments:
> 
> * It would good that the commit message explains what the commit does. This is unclear now.
I am not sure what you mean. This commit supports ANALYZE of an compressed hypertable. The commit message says so.
> * Is it possible that the number of uncompressed tuple will be incorrect after the chunk was compressed?
Only in 2.0, we added support to save number of tuples before compression. So it would be an estimate and not exact in other cases and/or if rows were deleted from the chunk.

Reply:
Then I suggest to check that `found_cnt == 0` and assert for `found_cnt == 1` after the error. Or keep current condition and assert for `found_cnt == 0` before the printing the error.

I am surprise that you implemented the loop for primary key access, instead of just reading singe value, which will be matching by definition.

Reply:
Maybe I am missing something here. But  how do you read without the iterator?

Reply:
I think the main concern is the loop, not the iterator itself. Either use the PG scan API directly, our scan API. We actually already have a `ts_scanner_scan_one` function. With the iterator we can do something similar `ts_scan_iterator_get_one(&iterator)`

Reply:
For instance, this could be added to `scan_iterator.h`:

```
static inline TupleInfo *
ts_scan_iterator_get_one(ScanIterator *iterator)
{
	TupleInfo *ti;
	
	ts_scanner_start_scan(&iterator->ctx, &iterator->ictx);
	ti = ts_scanner_next(&iterator->ctx, &iterator->ictx);
	
	if (ts_scanner_next(&iterator->ctx, &iterator->ictx) != NULL)
		elog(ERROR, "more than one tuple when one expected");
	
	return ti;
}
```